soundwire: stream: skip m_rt which is not in preparing stage

bardliao opened the pull request:
sdw_compute_group_params() will count payload bandwidth and it count all m_rt in the bus. The m_rt will be added to the bus by sdw_stream_add_master(). If 2 streams are opened simultaneously, the m_rt of the second stream will be counted when
sdw_compute_group_params() is counting the payload bandwidth of the first stream. Which is incorrect.
Set PREPARED state earlier so that sdw_compute_group_params() can skip the m_rt that is not in the preparing stage yet.

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -1456,15 +1464,15 @@ static int _sdw_prepare_stream(struct sdw_stream_runtime *stream,
 			dev_err(bus->dev, "Prepare port(s) failed ret = %d\n",
 				ret);
+			stream->state = state;
 			return ret;
```
I wonder why this is not jumping to `restore_params`?

bardliao replied:
We didn't change bus->params at this point yet.

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -1403,8 +1403,15 @@ static int _sdw_prepare_stream(struct sdw_stream_runtime *stream,
+	 * the stream that is CONFIGURED and PREPARED and skip the stream that is CONFIGURED
+	 * but not in the preparing stage.
+	 */
+	stream->state = SDW_STREAM_PREPARED;
```
I think in other patches we also moved the set to PREPARED state prior that we actually prepare things.

I got the reason and at the same time it sounds wrong to set something as PREPARED when it is not...

bardliao replied:
I agree that the stream->state may not represent the actual state perfectly. But, I can't find a better way to skip the m_rt that should not be counted yet. And stream->state is used by the stream helpers only and the helpers will be called one by one. So, I think it is safe to change the stream->state earlier in the same stream helper.

ranj063 replied:
@bardliao this really seems like a hack or a workaround. Maybe the best thing to do would be to revisit the state machine so that we can gracefully skip the runtime thats not ready to be included yet?

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -189,6 +189,9 @@ static int sdw_compute_group_params(struct sdw_bus *bus,
 	}
 
 	list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
+		if (m_rt->stream->state < SDW_STREAM_PREPARED)
```
I think the issue manifests because sdw_prepare_stream() looks at the m_rt associated with the stream but sdw_compute_group_params() looks at the bus->m_rt_list. Can you maybe filter this list in this function by passing the stream as an argument when called from sdw_prepare_stream()?

bardliao replied:
I thought about that. sdw_compute_group_params() lools at the bus->m_rt_list is because it needs to calculate all the active stream's bandwidth in the bus. We assume that all m_rt in the bus are active, but that is not true. And we did similar change in https://github.com/thesofproject/linux/commit/d7722a643421. That's why I choose to set the PREPARE state earlier.
But I am fine with the passing `stream` solution, too. PR updated.

bardliao committed:
SoundWire: pass stream to compute_params()

The stream parameter will be used in the follow up commit.
No function change.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
soundwire: generic_bandwidth_allocation: count the bandwidth of active streams only

sdw_compute_group_params() should only count payload bandwidth of the
active streams which is in the ENABLED and DISABLED state in the bus.
And add the payload bandwidth of the stream that calls
sdw_compute_group_params() in sdw_prepare_stream().

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ranj063 commented on include/linux/soundwire/sdw.h:
```diff
@@ -914,7 +969,7 @@ struct sdw_bus {
 	unsigned int link_id;
 	int id;
-	int (*compute_params)(struct sdw_bus *bus);
+	int (*compute_params)(struct sdw_bus *bus, struct sdw_stream_runtime *stream);
```
@bardliao for the same of keeping this patch simpler to read, can you move this struct definition instead?

bardliao replied:
Yes, done

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -190,6 +190,15 @@ static int sdw_compute_group_params(struct sdw_bus *bus,
 
 	list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
+		if (m_rt->stream == stream) {
+			if (stream->state != SDW_STREAM_CONFIGURED)
```
can this happen in practice?

bardliao replied:
Yes, bus->compute_params() -> sdw_compute_group_params() will be called in both _sdw_prepare_stream() and _sdw_deprepare_stream()

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -190,6 +190,15 @@ static int sdw_compute_group_params(struct sdw_bus *bus,
+			if (stream->state != SDW_STREAM_CONFIGURED)
+				continue;
+		} else {
+			if (m_rt->stream->state != SDW_STREAM_ENABLED &&
```
and this is not intuitive at all. What are you trying to say here? Ignore a runtime thats not associated with the stream if it has already been activated once?

bardliao replied:
No, to the opposite. Only calculate the bandwidth of the active streams, and add the bandwidth of the current stream.

ranj063 replied:
ok then the check for != ENABLED makes sense. but why also check for != DISABLED? Is this the case with open but paused streams? Anyway, I think this needs a comment here with the explanation. And the commit message needs to be modified too based on whatever explanation you add here.

bardliao replied:
Yes, DISABLED is about paused. Please see https://github.com/thesofproject/linux/blob/topic/sof-dev/Documentation/driver-api/soundwire/stream.rst#soundwire-stream-states and https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sdw_utils/soc_sdw_utils.c#L809

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -190,6 +190,15 @@ static int sdw_compute_group_params(struct sdw_bus *bus,
+		if (m_rt->stream == stream) {
+			if (stream->state != SDW_STREAM_CONFIGURED)
+				continue;
+		} else {
```
no need for an else after a continue btw

bardliao replied:
> no need for an else after a continue btw

We still need an else because it will not continue if m_rt->stream == stream and stream->state == SDW_STREAM_CONFIGURED and it will then continue because m_rt->stream->state != SDW_STREAM_ENABLED && m_rt->stream->state != SDW_STREAM_DISABLED.
We only want to check m_rt->stream->state != SDW_STREAM_ENABLED && m_rt->stream->state != SDW_STREAM_DISABLED when the m_rt->stream != stream.

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -190,6 +190,21 @@ static int sdw_compute_group_params(struct sdw_bus *bus,
 	list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
+		/*
+		 * sdw_compute_group_params could be called in the deprepared case, skip the
+		 * m_rt in that case.
```
a better comment would be to include the runtime only during prepare

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -190,6 +190,21 @@ static int sdw_compute_group_params(struct sdw_bus *bus,
+				continue;
+		/*
+		 * Stream stage ENABLED and DISABLED are between PREPARED and DEPREPARED,
+		 * We should count the bandwidth only for ENABLED and DISABLED stages.
```
@bardliao I think looking at the enum definition, its easy to see this. A better explanation would be something like "Include runtimes with running (ENABLED state) and paused (DISABLED state) streams"

bardliao replied:
Both updated.

ranj063 commented:
@bardliao there's a build error in qcom.c. You'll need to fix it too because you changed the signature

ujfalusi reviewed (changes requested):
@bardliao, I will block the merge to let you check the CI device fails on all the SDW config..

bardliao commented:
SOFCI TEST

ujfalusi reviewed (approved):
@bardliao, I guess the LNL fails are the usual ones, they look that way.
